### DISCUSSION TITLE
isDirty not changing when using defaultValues inside useForm

### DISCUSSION THREAD

**Discussion by keanan** [react-hook-form/react-hook-form#11567] [Q&A] [answered]
Hello,

I have tried to follow the examples and not sure what the issue is but I am using defaultValues inside useForm and isDirty does not change when using defaultValues. Please could someone assist if possible, thanks.

```
export default function App() {
  const {
    register,
    handleSubmit,
    formState: { isDirty, isValid },
  } = useForm({
    defaultValues: {
      firstName: "firstName",
    },
  });
  const onSubmit = (data) => console.log(data);

  console.log(isDirty, isValid);

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmit)}>
        <input {...register("firstName")} placeholder="First Name" />
         <input disabled={!isDirty} type="submit" />
      </form>
    </div>
  );
}
```

[Sandbox URL](https://codesandbox.io/p/sandbox/useformstate-forked-vym3d7?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt6zsbmj00063b6bwz816qtj%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt6zsbmj00023b6bqwmz8o7u%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt6zsbmj00033b6bsximaguj%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt6zsbmj00053b6bc1t9057x%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt6zsbmj00023b6bqwmz8o7u%2522%253A%257B%2522id%2522%253A%2522clt6zsbmj00023b6bqwmz8o7u%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt6ztqfk004p3b6bzm2t6euo%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt6ztqfk004p3b6bzm2t6euo%2522%257D%252C%2522clt6zsbmj00053b6bc1t9057x%2522%253A%257B%2522id%2522%253A%2522clt6zsbmj00053b6bc1t9057x%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt6zsbmj00043b6bgfoebb4w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt6zsbmj00043b6bgfoebb4w%2522%257D%252C%2522clt6zsbmj00033b6bsximaguj%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clt6zsbmj00033b6bsximaguj%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by sujinleeme** [comment]
@keanan The codesandbox link is broken or it seems it's not on public. Shall you update the url again?

**Reply by keanan** [reply]
Thanks @sujinleeme I have updated the URL.

**Reply by sujinleeme** [reply]
The `<input disabled={!isDirty} type="submit" />` works well as expected.

![image](https://github.com/react-hook-form/react-hook-form/assets/6011401/469e8bf8-ce37-480c-8f50-5d82c4cbe1b7)

You may forget to add the CSS `:disabled` selector of the submit button.

```css
input:disabled {
  opacity: 0.4;
}
```

**Reply by keanan** [reply]
Hi @sujinleeme thanks, but the issue I am facing is that `isDirty` should be `true` when `defaultValues` are supplied with `useForm`. Currently it does not change state so the submit button is always `disabled` even when all form inputs are pre-populated.

**Reply by sujinleeme** [reply, marked as answer]
Only providing `defaultValue` won't change `isDirty` because a comparison between defaultValue and current formValues will be made. 
To update the formValues at first, you must use `setValue`  or `reset`. 

```ts
  const {
    reset,
  } = useForm({
    defaultValues: {
      firstName: "",  // The value shouldn't be `test`. defaultValues can be omitted. 
    },
    mode: "onChange",
  });
  
  useEffect(() => {
    reset(
      {
        firstName: "test",
      },
      { keepDefaultValues: true }
    );
  }, [reset]);
```

Here is the [codesandbox](https://codesandbox.io/p/sandbox/11567-zygqcd?file=%2Fsrc%2FApp.js%3A16%2C1).

**Reply by keanan** [reply]
That is good to know, thanks for your help @sujinleeme 🙇‍♂️
